File: src/Day23.hs
module Day23 (solve1, solve2) where

import Control.Parallel.Strategies (parMap, rdeepseq)
import Data.Hashable (hashWithSalt)
import Data.Sequence qualified as Seq
import Data.Vector.Unboxed qualified as VU
import Data.Vector.Unboxed.Deriving (derivingUnbox)
import RIO
import RIO.HashMap qualified as HM
import RIO.HashMap.Partial qualified as HM'
import RIO.HashSet qualified as HS
import RIO.List qualified as L
import RIO.List.Partial (head)
import RIO.Text qualified as T
import Safe (maximumDef)
import Utils
import Prelude (print)

data Item = Path | Forest | UpSlope | DownSlope | LeftSlope | RightSlope deriving (Show, Eq)

itemToInt :: Item -> Int
itemToInt Path = 0
itemToInt Forest = 1
itemToInt UpSlope = 2
itemToInt DownSlope = 3
itemToInt LeftSlope = 4
itemToInt RightSlope = 5

intToItem :: Int -> Item
intToItem 0 = Path
intToItem 1 = Forest
intToItem 2 = UpSlope
intToItem 3 = DownSlope
intToItem 4 = LeftSlope
intToItem _ = RightSlope

derivingUnbox
  "Item"
  [t|Item -> Int|]
  [|itemToInt|]
  [|intToItem|]

data Direction = Up | Dow | Lef | Righ deriving (Show, Eq, Ord)

data Matrix = Matrix
  { _matrix :: !(VU.Vector Item),
    _width :: !Int,
    _height :: !Int
  }
  deriving (Show, Eq)

makeMatrix :: [[Item]] -> Matrix
makeMatrix xs =
  let width = length $ head xs
      height = length xs
      matrix = VU.fromList $ concat xs
   in Matrix matrix width height

getMatrixElement :: Matrix -> (Int, Int) -> Item
getMatrixElement (Matrix matrix width _) (x, y) = matrix VU.! (x * width + y)

instance Hashable Item where
  hashWithSalt salt Path = hashWithSalt salt (0 :: Int)
  hashWithSalt salt Forest = hashWithSalt salt (1 :: Int)
  hashWithSalt salt UpSlope = hashWithSalt salt (2 :: Int)
  hashWithSalt salt DownSlope = hashWithSalt salt (3 :: Int)
  hashWithSalt salt LeftSlope = hashWithSalt salt (4 :: Int)
  hashWithSalt salt RightSlope = hashWithSalt salt (5 :: Int)

towards :: Direction -> (Int, Int) -> (Int, Int)
towards Up (x, y) = (x - 1, y)
towards Dow (x, y) = (x + 1, y)
towards Lef (x, y) = (x, y - 1)
towards Righ (x, y) = (x, y + 1)

isFeasible :: Matrix -> (Int, Int) -> Bool
isFeasible matrix (x, y) = x >= 0 && y >= 0 && x < _height matrix && y < _width matrix && getMatrixElement matrix (x, y) /= Forest

getNearby :: (Int, Int) -> Matrix -> [(Int, Int)]
getNearby (x, y) matrix =
  let dirs = [Up, Dow, Lef, Righ]
      goto = map (\i -> towards i (x, y)) dirs
   in filter (isFeasible matrix) goto

isJunction :: (Int, Int) -> Matrix -> Bool
isJunction (x, y) matrix = length (getNearby (x, y) matrix) > 2 || x == 0 || x == _height matrix - 1

-- Returns: (x, y), direction, steps (weight)
towardsUntilJunction :: Direction -> (Int, Int) -> Matrix -> Int -> Maybe ((Int, Int), Direction, Int)
towardsUntilJunction dir (x, y) matrix acc =
  let dirs' =
        [dir] <> case dir of
          Up -> [Lef, Righ]
          Dow -> [Lef, Righ]
          Lef -> [Up, Dow]
          Righ -> [Up, Dow]
      goto_ = filter (\(_, j) -> isFeasible matrix j) $ map (\i -> (i, towards i (x, y))) dirs'
      (dir_, (x_, y_)) = head goto_
   in case length goto_ of
        0 -> Nothing
        _ -> if isJunction (x_, y_) matrix then Just ((x_, y_), dir_, acc + 1) else towardsUntilJunction dir_ (x_, y_) matrix (acc + 1)

parse :: Text -> Matrix
parse input =
  let chars = (map T.unpack $ T.lines input)
   in makeMatrix
        ( map
            ( map
                ( \case
                    '#' -> Forest
                    '.' -> Path
                    '^' -> UpSlope
                    'v' -> DownSlope
                    '<' -> LeftSlope
                    '>' -> RightSlope
                    _ -> error "unexpected char"
                )
            )
            chars
        )

-- old naive dfs
dfs :: Matrix -> (Int, Int) -> Bool -> Int
dfs matrix (startx, starty) ignoreSlope =
  go (startx, starty) HS.empty 0
  where
    go :: (Int, Int) -> HS.HashSet (Int, Int) -> Int -> Int
    go (x, y) hs acc =
      let dirs =
            -- ignoreSlope is always false though...
            if not ignoreSlope
              then case getMatrixElement matrix (x, y) of
                UpSlope -> [Up]
                DownSlope -> [Dow]
                LeftSlope -> [Lef]
                RightSlope -> [Righ]
                _ -> [Up, Dow, Lef, Righ]
              else [Up, Dow, Lef, Righ]
          goto = map (`towards` (x, y)) dirs
          goto' = filter (\i -> not $ HS.member i hs) $ filter (isFeasible matrix) goto
          hs' = HS.insert (x, y) hs
       in if x == _height matrix - 1
            then acc
            else case goto' of
              -- dead end, this shall not be counted in
              [] -> -1
              _ ->
                foldl'
                  ( \acc' i ->
                      let res = go i hs' (acc + 1)
                          maxAcc = if res == -1 then acc' else max acc' res
                       in maxAcc
                  )
                  0
                  goto'

type Graph = HashMap (Int, Int) [((Int, Int), Int)]

makeJunctionGraph :: (Int, Int) -> Matrix -> Graph
makeJunctionGraph start matrix =
  let bfs :: Seq.Seq ((Int, Int), Direction) -> HS.HashSet (Int, Int) -> Graph -> (HS.HashSet (Int, Int), Graph)
      bfs q h g =
        case Seq.viewl q of
          Seq.EmptyL -> (h, g)
          ((x, y), dir) Seq.:< xs ->
            let res = towardsUntilJunction dir (x, y) matrix 0
                h' = HS.insert (x, y) h
                nextDirs d = case d of
                  Up -> [Lef, Righ, Up]
                  Dow -> [Lef, Righ, Dow]
                  Lef -> [Up, Dow, Lef]
                  Righ -> [Up, Dow, Righ]
             in case res of
                  Nothing -> bfs xs h' g
                  Just ((x', y'), dir', weight) ->
                    if HS.member (x', y') h'
                      then bfs xs h' g
                      else
                        let g' = HM.insertWith (++) (x', y') [((x, y), weight)] $ HM.insertWith (++) (x, y) [((x', y'), weight)] g
                            g'' = HM.insert (x', y') (L.nub $ g' HM'.! (x', y')) $ HM.insert (x, y) (L.nub $ g' HM'.! (x, y)) g'
                            q' = xs <> Seq.fromList (map ((x', y'),) (nextDirs dir'))
                         in bfs q' h' g''
   in snd $ bfs (Seq.singleton (start, Dow)) HS.empty HM.empty

parallelArg :: Int
parallelArg = 4

-- new dfs with junction graph
dfs' :: Graph -> (Int, Int) -> (Int -> Bool) -> Int
dfs' graph start isEnd =
  go start HS.empty 0 0
  where
    go :: (Int, Int) -> HS.HashSet (Int, Int) -> Int -> Int -> Int
    go (x, y) hs acc depth =
      let goto = HM.lookupDefault [] (x, y) graph
          goto' = filter (\i -> not $ HS.member (fst i) hs) goto
          hs' = HS.insert (x, y) hs
          l =
            if depth < parallelArg
              then parMap rdeepseq (\i -> go (fst i) hs' (acc + snd i) (depth + 1)) goto'
              else map (\i -> go (fst i) hs' (acc + snd i) (depth + 1)) goto'
       in if isEnd x
            then acc
            else case goto' of
              -- dead end, this shall not be counted in
              [] -> -1
              _ -> l & filter (/= -1) & maximumDef 0

solve1 :: Text -> IO ()
solve1 input =
  let matrix = parse input
      start_ = [(0, y) | y <- [0 .. _width matrix - 1], getMatrixElement matrix (0, y) /= Forest]
      start = head $ assert' (length start_ == 1) start_
   in print $ dfs matrix start False

solve2 :: Text -> IO ()
solve2 input =
  let matrix = parse input
      start_ = [(0, y) | y <- [0 .. _width matrix - 1], getMatrixElement matrix (0, y) /= Forest]
      start = head $ assert' (length start_ == 1) start_
      junctionGraph = makeJunctionGraph start matrix
   in print $ dfs' junctionGraph start (\x -> x == _height matrix - 1)

File: comparison/day23.rs
use std::{
    collections::{HashMap, HashSet, VecDeque},
    fs::File,
    io::{BufRead, BufReader},
};

#[derive(Debug, PartialEq, Clone, Copy)]
enum Item {
    Path,
    Forest,
    UpSlope,
    DownSlope,
    LeftSlope,
    RightSlope,
}

#[derive(Debug)]
struct Matrix {
    data: Vec<Item>,
    width: i64,
    height: i64,
}

impl Matrix {
    fn get(&self, x: i64, y: i64) -> &Item {
        self.data.get((x * self.width + y) as usize).unwrap()
    }

    fn feasible(&self, x: i64, y: i64) -> bool {
        x >= 0 && y >= 0 && x < self.height && y < self.width && self.get(x, y) != &Item::Forest
    }
}

#[derive(Debug, Clone, Copy)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

fn towards(direction: &Direction, x: i64, y: i64) -> (i64, i64) {
    match direction {
        Direction::Up => (x - 1, y),
        Direction::Down => (x + 1, y),
        Direction::Left => (x, y - 1),
        Direction::Right => (x, y + 1),
    }
}

fn get_nearby(matrix: &Matrix, x: i64, y: i64) -> Vec<(i64, i64)> {
    let mut result = Vec::new();
    for direction in vec![
        Direction::Up,
        Direction::Down,
        Direction::Left,
        Direction::Right,
    ] {
        let (x, y) = towards(&direction, x, y);
        if matrix.feasible(x, y) {
            result.push((x, y));
        }
    }
    result
}

fn is_junction(matrix: &Matrix, x: i64, y: i64) -> bool {
    let count = get_nearby(matrix, x, y).len();
    count > 2 || x == 0 || x == matrix.height - 1
}

fn towards_until_junction(
    matrix: &Matrix,
    x: i64,
    y: i64,
    direction: &Direction,
    acc: i64,
) -> Option<((i64, i64), Direction, i64)> {
    let mut dirs = vec![direction.clone()];
    match direction {
        Direction::Up => dirs.extend(vec![Direction::Left, Direction::Right]),
        Direction::Down => dirs.extend(vec![Direction::Left, Direction::Right]),
        Direction::Left => dirs.extend(vec![Direction::Up, Direction::Down]),
        Direction::Right => dirs.extend(vec![Direction::Up, Direction::Down]),
    }
    for dir in dirs {
        let (xnew, ynew) = towards(&dir, x, y);
        if matrix.feasible(xnew, ynew) {
            if is_junction(matrix, xnew, ynew) {
                return Some(((xnew, ynew), dir, acc + 1));
            } else {
                return towards_until_junction(matrix, xnew, ynew, &dir, acc + 1);
            }
        }
    }
    None
}

type Graph = HashMap<(i64, i64), Vec<((i64, i64), i64)>>;

fn on2_dedup(v: &Vec<((i64, i64), i64)>) -> Vec<((i64, i64), i64)> {
    let mut result = Vec::new();
    for (xy, weight) in v {
        if !result.iter().any(|(xy_, _)| xy_ == xy) {
            result.push((*xy, *weight));
        }
    }
    result
}

fn make_junction_graph(matrix: &Matrix, x: i64, y: i64) -> Graph {
    fn next_dirs(d: &Direction) -> Vec<Direction> {
        match d {
            Direction::Up => vec![Direction::Left, Direction::Right, Direction::Up],
            Direction::Down => vec![Direction::Left, Direction::Right, Direction::Down],
            Direction::Left => vec![Direction::Up, Direction::Down, Direction::Left],
            Direction::Right => vec![Direction::Up, Direction::Down, Direction::Right],
        }
    }
    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();
    let mut graph: Graph = HashMap::new();
    queue.push_back(((x, y), Direction::Down));
    while !queue.is_empty() {
        let ((x, y), direction) = queue.pop_front().unwrap();
        visited.insert((x, y));
        let res = towards_until_junction(matrix, x, y, &direction, 0);
        let ((x_, y_), dir_, weight) = match res {
            None => continue,
            Some(res) => res,
        };
        if visited.contains(&(x_, y_)) {
            continue;
        }
        graph
            .entry((x, y))
            .or_insert(Vec::new())
            .push(((x_, y_), weight));
        graph
            .entry((x_, y_))
            .or_insert(Vec::new())
            .push(((x, y), weight));
        // dedup
        let gxy = graph.get(&(x, y)).unwrap();
        // gxy.dedup();
        graph.insert((x, y), on2_dedup(gxy));
        let gxy_ = graph.get(&(x_, y_)).unwrap();
        // gxy_.dedup();
        graph.insert((x_, y_), on2_dedup(gxy_));
        for d in next_dirs(&dir_) {
            queue.push_back(((x_, y_), d));
        }
    }
    graph
}

fn dfs<F>(
    graph: &Graph,
    visited: &mut HashSet<(i64, i64)>,
    x: i64,
    y: i64,
    acc: i64,
    is_end: &F,
) -> i64
where
    F: Fn(i64) -> bool,
{
    if is_end(x) {
        return acc;
    }
    let nearby = match graph.get(&(x, y)) {
        None => return -1,
        Some(nearby) => nearby,
    };
    let nearby: Vec<_> = nearby
        .iter()
        .filter(|(xy, _)| !visited.contains(xy))
        .collect();
    if nearby.is_empty() {
        return -1;
    }
    let mut max = -1;
    for (xy, weight) in nearby {
        visited.insert(*xy);
        let res = dfs(graph, visited, xy.0, xy.1, acc + weight, is_end);
        if res != -1 && res > max {
            max = res;
        }
        visited.remove(xy);
    }
    max
}

fn main() {
    let file = File::open("../input/day23").unwrap();
    let reader = BufReader::new(file).lines();
    let mut matrix = Matrix {
        data: Vec::new(),
        width: 0,
        height: 0,
    };
    for line in reader {
        let line = line.unwrap();
        assert!(matrix.width == 0 || matrix.width as usize == line.len());
        matrix.width = line.len() as i64;
        matrix.height += 1;
        for c in line.chars() {
            matrix.data.push(match c {
                '.' => Item::Path,
                '#' => Item::Forest,
                '^' => Item::UpSlope,
                'v' => Item::DownSlope,
                '<' => Item::LeftSlope,
                '>' => Item::RightSlope,
                _ => panic!("Unknown character"),
            });
        }
    }
    let mut start: (i64, i64) = (i64::MAX, i64::MAX);
    for y in 0..matrix.width {
        if matrix.get(0, y) != &Item::Forest {
            start = (0, y);
            break;
        }
    }
    assert!(start != (i64::MAX, i64::MAX));
    let junction_graph = make_junction_graph(&matrix, start.0, start.1);
    println!("{:?} {:?}", start, junction_graph);
    let mut visited = HashSet::new();
    visited.insert(start);
    let res = dfs(&junction_graph, &mut visited, start.0, start.1, 0, &|x| {
        x == matrix.height - 1
    });
    println!("Part 2: {}", res);
}
